### ISSUE TITLE
Any way to share depth texture on different render passes?

### ISSUE THREAD

**Issue by geroge-d** [google/filament#2138]
Hi guys 
  We want to render a large scene with thousands of models on mobile , so if we render it in one pass, it will be out of memory. We want to render it in multiple passes, like below, but the filament automatically clears the depth buffer ,despite we call setClearTarget( false,false,false) of view before rendering.
So could you guys please give me any clue to implement this process? thanks in adavance
Codes:
if ( engine.beginFrame(...))
{
     do some textures setup()
     render.render(view);
     do some other textures setup()
     view.setClearTarget(false,false,false);
     render.render(view)   //it still clears depth buffer anyway
}

**Comment by geroge-d** [comment]
We found related codes  below in render.cpp,comment the filament original codes replaced by  TargetBufferFlags clearFlags = TargetBufferFlags(viewClearFlags);   BUT it still doesn't work. And then , we found another code snipet in Framegraph ,but don't figure out how to implement it 
Code in Render.cpp:
  // We only honor the view's color buffer clear flags, depth/stencil are handled by the FrameGraph
    uint8_t viewClearFlags = view.getClearFlags() & (uint8_t)TargetBufferFlags::ALL;

    // FIXME: when the view doesn't ask for a clear, but it's drawn in an intermediate buffer
    //        that buffer needs to be cleared with transparent pixels if blending is enabled
//    TargetBufferFlags clearFlags = (TargetBufferFlags(viewClearFlags) & TargetBufferFlags::COLOR)
//                                   | TargetBufferFlags::DEPTH;
    TargetBufferFlags clearFlags = TargetBufferFlags(viewClearFlags);
Codes in FrameGraph.cpp:
  // TODO: make this better
            //  When replacing a resource from an imported render-target, we find all existing
            //  render-target that used that resource as the color attachment, and we clear
            //  all other attachments -- we have to do this because the resource aliasing
            //  must also alias the render targets.
            //  For instance, if a render-target was declared with a color+depth attachment,
            //  and the color attachment was aliased with an imported render-target, the
            //  render-target cache (see: equals()) wouldn't match it (because the depth
            //  attachment would be missing).
            for (fg::RenderTarget& rt : renderTargets) {
                auto& textures = rt.desc.attachments.textures;
                if (textures[0].isValid()) {
                    FrameGraphHandle handle = textures[0];
                    ResourceNode const& node = resourceNodes[handle.index];
                    if (node.resource->imported && node.resource == from.resource) {
                        for (size_t i = 1; i < textures.size(); ++i) {
                            textures[i] = {};
                        }
                    }
                }
            }

**Comment by romainguy** [comment]
@pixelflinger

**Comment by pixelflinger** [comment]
In addition to call `setClearTarget`, you also need to call  `setRenderTarget` to set the discard flags properly:
Like this: `view->setRenderTarget(nullptr, TargetBufferFlags::NONE);`

Please let us know if this helps.

**Comment by geroge-d** [comment]
> In addition to call `setClearTarget`, you also need to call `setRenderTarget` to set the discard flags properly:
> Like this: `view->setRenderTarget(nullptr, TargetBufferFlags::NONE);`
> 
> Please let us know if this helps.

i have aready tried this method ,and export method to java. it doesn't work

**Comment by geroge-d** [comment]
in source codes, we found filament clear user -set depth clear flag. @pixelflinger 
TargetBufferFlags clearFlags = (TargetBufferFlags(viewClearFlags) & TargetBufferFlags::COLOR)
 | TargetBufferFlags::DEPTH;

**Comment by pixelflinger** [comment]
I think this might be a bug in filament.  I will look into it but it might take a few days.

**Comment by geroge-d** [comment]
> I think this might be a bug in filament. I will look into it but it might take a few days.

Thanks.  We expect this fix to update our app

**Comment by pixelflinger** [comment]
Okay so looking at this more, it turns out that we don't support this use case at all. This is because the depth buffer is not retained between Views, because it only exists in the internal framegraph.
When drawing several Views, they're not fact drawn into the same framebuffer -- they're just blended in. In other words, Views are totally independent.

There is one exception to this though, which *may* work for you. If you disable post-processing, then in that case, we are rendering everything into the same framebuffer. However, due to the line you highlighted, this actually doesn't work currently -- but that's a bug, and I will fix it.

So in short, your scenario can only work if you disable post-processing and once we fix that problem where we always clear the depth buffer. I'm not sure if disabling post-processing is viable for you?

Now I want to know, what runs out of memory when you render thousand of objects? is it filament? you could increase the size of the command buffer for instance.

**Comment by geroge-d** [comment]
Thanks for your reply @pixelflinger 
First, we only render onto the same view
Second,  when render a very large scene, the gpu memory runs OOM. So we want this rendering pass be split into two steps, which renders a part of scene
Third, We will give a try to disable post-processing

**Comment by pixelflinger** [comment]
Can you tell me more about the OOM, I would like to understand that. What's the actual error?

**Comment by geroge-d** [comment]
the issue 2057 ,we have reproduced it before,and we fix it by increasing this to a 9M buffer size.
Now the oom we met ,is caused by rendering a very large scene, the gpu cannot allocate more memory to render the models . We saw GPU memory footprint is about over 1 GB in Android studio.  So, we want to split the orginal render pass into two or more render pass, in order to fix the oom issue.

**Comment by pixelflinger** [comment]
You're definitely outside of the use cases we had in mind!  Okay, in that case, your only option right now, is to disable post-processing and to wait for a fix for the depth buffer clear (hopefully shortly).

In the long term, a better solution would be for us to handle this -- i.e. split the scene rendering into several render passes (I'm assuming this is the problem here).

Are you able to tell us how many objects you're rendering and how many vertices in total?

**Comment by geroge-d** [comment]
our app is a video editor with function 3d modeling. sometimes artists will make a video with about   100-500 layers , which have about hundred or thousand of vertices and a lot of textures. So in this situation , the filament will crash if we render it in one pass

**Comment by romainguy** [comment]
Are you running out of memory in filament's command buffer? Or is it gpu memory that's an issue? If the former there's a much simpler fix.

**Comment by geroge-d** [comment]
> Are you running out of memory in filament's command buffer? Or is it gpu memory that's an issue? If the former there's a much simpler fix.

No , it is not caused by the command buffer  . the gpu memory is the issue
